build-support/bin/native/bootstrap_code.sh

```diff
@@ -0,0 +1,106 @@
+readonly MODE="${MODE:-release}"
+case "$MODE" in
+  debug) MODE_FLAG="" ;;
+  *) MODE_FLAG="--release" ;;
+esac
+
+readonly NATIVE_ENGINE_CACHE_DIR=${CACHE_ROOT}/bin/native-engine
+
+function calculate_current_hash() {
+  # Cached and unstaged files, with ignored files excluded.
+  # NB: We fork a subshell because one or both of `ls-files`/`hash-object` are
+  # sensitive to the CWD, and the `--work-tree` option doesn't seem to resolve that.
+  (
+   cd ${REPO_ROOT}
+   (echo "${MODE_FLAG}"
+    echo "${RUST_TOOLCHAIN}"
+    git ls-files -c -o --exclude-standard \
+     "${NATIVE_ROOT}" \
+     "${REPO_ROOT}/src/python/pants/engine/native.py" \
+     "${REPO_ROOT}/build-support/bin/native" \
+     "${REPO_ROOT}/3rdparty/python/requirements.txt" \
+   | grep -v -E -e "/BUILD$" -e "/[^/]*\.md$" \
+   | git hash-object -t blob --stdin-paths) | fingerprint_data
+  )
+}
+
+function _build_native_code() {
+  # Builds the native code, and echos the path of the built binary.
+
+  "${REPO_ROOT}/build-support/bin/native/cargo" build ${MODE_FLAG} \
+    --manifest-path ${NATIVE_ROOT}/Cargo.toml || die
+  echo "${NATIVE_ROOT}/target/${MODE}/libengine.${LIB_EXTENSION}"
+}
+
+function bootstrap_native_code() {
+  # Bootstraps the native code only if needed.
+  local native_engine_version="$(calculate_current_hash)"
+  local engine_version_header="engine_version: ${native_engine_version}"
+  local target_binary="${NATIVE_ENGINE_CACHE_DIR}/${native_engine_version}/${NATIVE_ENGINE_BINARY}"
+  local target_binary_metadata="${target_binary}.metadata"
+  if [[ ! -f "${target_binary}" || ! -f "${target_binary_metadata}" ]]
+  then
+    local -r native_binary="$(_build_native_code)"
+
+    # If bootstrapping the native engine fails, don't attempt to run pants
+    # afterwards.
+    if ! [ -f "${native_binary}" ]
+    then
+      die "Failed to build native engine."
+    fi
+
+    # Pick up Cargo.lock changes if any caused by the `cargo build`.
+    native_engine_version="$(calculate_current_hash)"
+    engine_version_header="engine_version: ${native_engine_version}"
+    target_binary="${NATIVE_ENGINE_CACHE_DIR}/${native_engine_version}/${NATIVE_ENGINE_BINARY}"
+    target_binary_metadata="${target_binary}.metadata"
+
+    mkdir -p "$(dirname ${target_binary})"
+    cp "${native_binary}" "${target_binary}"
+
+    local -r metadata_file=$(mktemp -t pants.native_engine.metadata.XXXXXX)
+    echo "${engine_version_header}" > "${metadata_file}"
+    echo "repo_version: $(git describe --dirty)" >> "${metadata_file}"
+    mv "${metadata_file}" "${target_binary_metadata}"
+  fi
+
+  # Establishes the native engine wheel resource only if needed.
+  # NB: The header manipulation code here must be coordinated with header stripping code in
+  #     the Native.binary method in src/python/pants/engine/native.py.
+  if [[
+    ! -f "${NATIVE_ENGINE_RESOURCE}" ||
+    "$(head -1 "${NATIVE_ENGINE_RESOURCE}" | tr '\0' '\n')" != "${engine_version_header}"
```

Comment:
Can you throw in a `2>/dev/null` to the `tr` command? I'm getting logspam from it when building because of non-unicode byte sequences...

Reply:
Done.